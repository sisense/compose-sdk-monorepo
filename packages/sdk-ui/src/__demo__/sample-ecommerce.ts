import { measureFactory } from '@sisense/sdk-data';

import { Commerce, Brand, Country, Category, DataSource } from './sample-ecommerce-autogenerated';

export { Commerce, Brand, Category, Country, DataSource };

/** adding predefined set of measures */
export const Measures = {
  SumCost: measureFactory.sum(Commerce.Cost).format('0,0$'),
  AvgCost: measureFactory.average(Commerce.Cost).format('0,0$'),
  SumRevenue: measureFactory.sum(Commerce.Revenue, 'Total Revenue').format('0,0$'),
  Quantity: measureFactory.sum(Commerce.Quantity, 'Total Quantity').format('0,0'),
  CountBrands: measureFactory.countDistinct(Commerce.BrandID).format('0,0'),
};
