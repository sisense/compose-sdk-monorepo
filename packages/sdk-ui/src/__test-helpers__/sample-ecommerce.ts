import {
  Dimension,
  DateDimension,
  Attribute,
  createAttribute,
  createDateDimension,
  createDimension,
} from '@sisense/sdk-data';

export const DataSource = 'Sample ECommerce';

interface BrandDimension extends Dimension {
  Brand: Attribute;
  BrandID: Attribute;
}
export const Brand = createDimension({
  name: 'Brand',
  Brand: createAttribute({
    name: 'Brand',
    type: 'text-attribute',
    expression: '[Brand.Brand]',
  }),
  BrandID: createAttribute({
    name: 'BrandID',
    type: 'numeric-attribute',
    expression: '[Brand.Brand ID]',
  }),
}) as BrandDimension;

interface CategoryDimension extends Dimension {
  Category: Attribute;
  CategoryID: Attribute;
}
export const Category = createDimension({
  name: 'Category',
  Category: createAttribute({
    name: 'Category',
    type: 'text-attribute',
    expression: '[Category.Category]',
  }),
  CategoryID: createAttribute({
    name: 'CategoryID',
    type: 'numeric-attribute',
    expression: '[Category.Category ID]',
  }),
}) as CategoryDimension;

interface CommerceDimension extends Dimension {
  AgeRange: Attribute;
  BrandID: Attribute;
  CategoryID: Attribute;
  Condition: Attribute;
  Cost: Attribute;
  CountryID: Attribute;
  DateMonth: Attribute;
  Gender: Attribute;
  Quantity: Attribute;
  Revenue: Attribute;
  VisitID: Attribute;
  Date: DateDimension;
}
export const Commerce = createDimension({
  name: 'Commerce',
  AgeRange: createAttribute({
    name: 'AgeRange',
    type: 'text-attribute',
    expression: '[Commerce.Age Range]',
  }),
  BrandID: createAttribute({
    name: 'BrandID',
    type: 'numeric-attribute',
    expression: '[Commerce.Brand ID]',
  }),
  CategoryID: createAttribute({
    name: 'CategoryID',
    type: 'numeric-attribute',
    expression: '[Commerce.Category ID]',
  }),
  Condition: createAttribute({
    name: 'Condition',
    type: 'text-attribute',
    expression: '[Commerce.Condition]',
  }),
  Cost: createAttribute({
    name: 'Cost',
    type: 'numeric-attribute',
    expression: '[Commerce.Cost]',
  }),
  CountryID: createAttribute({
    name: 'CountryID',
    type: 'numeric-attribute',
    expression: '[Commerce.Country ID]',
  }),
  DateMonth: createAttribute({
    name: 'DateMonth',
    type: 'numeric-attribute',
    expression: '[Commerce.Date (Month)]',
  }),
  Gender: createAttribute({
    name: 'Gender',
    type: 'text-attribute',
    expression: '[Commerce.Gender]',
  }),
  Quantity: createAttribute({
    name: 'Quantity',
    type: 'numeric-attribute',
    expression: '[Commerce.Quantity]',
  }),
  Revenue: createAttribute({
    name: 'Revenue',
    type: 'numeric-attribute',
    expression: '[Commerce.Revenue]',
  }),
  VisitID: createAttribute({
    name: 'VisitID',
    type: 'numeric-attribute',
    expression: '[Commerce.Visit ID]',
  }),
  Date: createDateDimension({
    name: 'Date',
    expression: '[Commerce.Date (Calendar)]',
  }),
}) as CommerceDimension;

interface CountryDimension extends Dimension {
  Country: Attribute;
  CountryID: Attribute;
}
export const Country = createDimension({
  name: 'Country',
  Country: createAttribute({
    name: 'Country',
    type: 'text-attribute',
    expression: '[Country.Country]',
  }),
  CountryID: createAttribute({
    name: 'CountryID',
    type: 'numeric-attribute',
    expression: '[Country.Country ID]',
  }),
}) as CountryDimension;
