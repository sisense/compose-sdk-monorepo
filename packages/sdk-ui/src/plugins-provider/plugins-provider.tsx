import { ReactNode, useCallback, useContext, useMemo, useRef } from 'react';
import { PluginsContext } from './plugins-context';
import { PluginComponent } from './types';

/**
 * Plugin Provider component that allows registering and accessing plugins.
 *
 * @internal
 */
export const PluginsProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const pluginMapRef = useRef(new Map<string, PluginComponent>());

  const registerPlugin = useCallback((pluginType: string, plugin: PluginComponent) => {
    if (!pluginMapRef.current.has(pluginType)) {
      pluginMapRef.current.set(pluginType, plugin);
    }
  }, []);

  const getPlugin = useCallback((pluginType: string) => {
    return pluginMapRef.current.get(pluginType);
  }, []);

  const plugins = useMemo(() => pluginMapRef.current, []);

  return (
    <PluginsContext.Provider value={{ registerPlugin, getPlugin, plugins }}>
      {children}
    </PluginsContext.Provider>
  );
};

/**
 * @internal
 */
export const usePlugins = () => useContext(PluginsContext);
