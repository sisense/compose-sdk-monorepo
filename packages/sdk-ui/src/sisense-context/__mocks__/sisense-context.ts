export const useSisenseContext = vi.fn();
export const useSisenseContextMock = useSisenseContext;
