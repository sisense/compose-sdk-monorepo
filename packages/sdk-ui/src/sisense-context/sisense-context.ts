import { createContext, useContext } from 'react';
import { ClientApplication } from '../app/client-application';

export type SisenseContextPayload = {
  isInitialized: boolean;
  app?: ClientApplication;
  tracking: {
    enabled: boolean;
    packageName: string;
  };
};

export const SisenseContext = createContext<SisenseContextPayload>({
  isInitialized: false,
  tracking: {
    enabled: true,
    packageName: 'sdk-ui',
  },
});

export const useSisenseContext = () => useContext(SisenseContext);
