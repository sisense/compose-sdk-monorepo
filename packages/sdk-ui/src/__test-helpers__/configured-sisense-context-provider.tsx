import { mockToken, mockUrl } from '@/__mocks__/msw';
import { SisenseContextProviderProps } from '@/props';
import { SisenseContextProvider } from '@/sisense-context/sisense-context-provider';

const contextProviderProps: SisenseContextProviderProps = {
  url: mockUrl,
  token: mockToken,
  defaultDataSource: 'Sample ECommerce',
  appConfig: {
    queryCacheConfig: { enabled: false },
    trackingConfig: { enabled: false },
  },
};

export const ConfiguredSisenseContextProvider: React.FC<{
  children: React.ReactNode;
}> = ({ children }) => {
  return <SisenseContextProvider {...contextProviderProps}>{children}</SisenseContextProvider>;
};
