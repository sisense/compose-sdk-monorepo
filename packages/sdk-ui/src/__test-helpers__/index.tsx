/**
 * Common test utilities
 */
import { PropsWithChildren, ReactElement } from 'react';
import { render, RenderResult } from '@testing-library/react';
import type { Cell, Data } from '@ethings-os/sdk-data';
import isObject from 'lodash-es/isObject';
import userEvent from '@testing-library/user-event';
import { SisenseContext, SisenseContextPayload } from '../sisense-context/sisense-context';
import { Authenticator, HttpClient } from '@ethings-os/sdk-rest-client';
import { ClientApplication } from '../app/client-application';
import { DeepPartial } from 'ts-essentials';
import { useTranslation } from 'react-i18next';
import { translation } from '../translation/resources/en';
import { CustomWidgetsProvider } from '../custom-widgets-provider';
import { MenuProvider } from '@/common/components/menu/menu-provider';

type UserSetupFn = (typeof userEvent)['setup'];
interface SetupResult extends RenderResult {
  user: ReturnType<UserSetupFn>;
}
/**
 * This helper method combines the userEvent.setup() and render() calls at the
 * start of a unit test using @testing-library/react.
 *
 * Based on recommendation by: https://testing-library.com/docs/user-event/intro/#writing-tests-with-userevent
 *
 * @param jsx - React component instance to test, passed directly to render()
 * @param useFakeTimers - true if Vitest fake timers are enabled
 * @returns
 */
export function setup(jsx: ReactElement, useFakeTimers = false): SetupResult {
  return {
    // Need to set "advanceTimers" if we're using Vitest's fake timers.
    // The solution in https://github.com/testing-library/user-event/issues/833#issuecomment-1171452841
    // is for Jest, but this seems to work for Vitest as well.
    user: userEvent.setup(
      useFakeTimers ? { advanceTimers: (ms) => vi.advanceTimersByTime(ms) } : undefined,
    ),
    ...render(jsx),
  };
}

export const MockedHighchartsReact = ({
  options,
}: {
  /**
   * Highcharts chart configuration object.
   * Please refer to the Highcharts (API documentation)[https://api.highcharts.com/highcharts/].
   */
  options: object;
}) => {
  return <div>{JSON.stringify(options)}</div>;
};

type MockedSisenseContextProviderProps = PropsWithChildren<
  DeepPartial<Omit<SisenseContextPayload, 'app'>>
>;

const mockHttpClient = new HttpClient(
  'http://mock-url/sometenant?someparam=true',
  {} as Authenticator,
  'mock-env',
);
export const MockedSisenseContextProvider = ({
  children,
  tracking = {
    enabled: true,
  },
  isInitialized = true,
  errorBoundary,
}: MockedSisenseContextProviderProps) => {
  return (
    <SisenseContext.Provider
      value={{
        app: {
          httpClient: mockHttpClient,
        } as ClientApplication,
        tracking: {
          enabled: tracking.enabled ?? true,
          packageName: tracking.packageName ?? 'sdk-ui',
        },
        isInitialized,
        errorBoundary: { showErrorBox: true, ...errorBoundary },
      }}
    >
      <MenuProvider>
        <CustomWidgetsProvider>{children}</CustomWidgetsProvider>
      </MenuProvider>
    </SisenseContext.Provider>
  );
};

/**
 * Extends the provided data by blurring specified rows.
 *
 * @param {Data} data - The original data object.
 * @param {number[]} blurRowsIndexes - An array of row indexes to be blurred.
 * @returns {Data} The data object with blurred rows.
 */
export const withBlurredRows = (data: Data, blurRowsIndexes: number[]) => ({
  ...data,
  rows: data.rows.map((row, rowIndex) =>
    row.map((value) => {
      const cell: Cell = isObject(value) ? value : { data: value };

      if (blurRowsIndexes.includes(rowIndex)) {
        return { ...cell, blur: true };
      }
      return cell;
    }),
  ),
});

/**
 * Mocking the i18next translation functionality for tests.
 */
export const setupI18nMock = () => {
  vi.mock('react-i18next', async () => {
    const original: { useTranslation: typeof useTranslation } = await vi.importActual(
      'react-i18next',
    );

    // Helper function to get nested translations safely
    const getNestedTranslation = (key: string): string | undefined => {
      return key.split('.').reduce((accumulator: any, currentKey: string) => {
        if (accumulator && typeof accumulator === 'object' && currentKey in accumulator) {
          return accumulator[currentKey];
        }
        return undefined;
      }, translation) as string | undefined;
    };

    return {
      ...original,
      useTranslation: () => ({
        ...original.useTranslation(),
        t: vi.fn((key, options?: Record<string, any>) => {
          const translatedValue = getNestedTranslation(key);
          if (translatedValue && options) {
            // Replace variables in the translated string
            return translatedValue.replace(/{{(.*?)}}/g, (_, varName) => {
              const trimmedVarName = varName.trim();
              return options[trimmedVarName] || `{{${trimmedVarName}}}`; // Fallback if variable is not found
            });
          }
          return translatedValue ?? key; // Fallback to the key if translation is not found
        }),
      }),
    };
  });
};
