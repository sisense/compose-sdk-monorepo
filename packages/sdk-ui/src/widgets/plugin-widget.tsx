import { DynamicSizeContainer, getWidgetDefaultSize } from '@/dynamic-size-container';
import ErrorBoundaryBox from '@/error-boundary/error-boundary-box';
import { usePlugins } from '@/plugins-provider';
import { PluginComponentProps } from '@/plugins-provider/types';
import { PluginWidgetProps } from '@/props';
import { useSisenseContext } from '@/sisense-context/sisense-context';
import { getDataSourceName } from '@sisense/sdk-data';
import { WidgetContainer } from './common/widget-container';
import { useTranslation } from 'react-i18next';
import { WidgetStyleOptions } from '@/types';

/**
 * Component that renders a plugin widget.
 *
 * @internal
 */
export const PluginWidget: React.FC<PluginWidgetProps> = (widgetProps) => {
  const { t } = useTranslation();
  const { plugins } = usePlugins();
  const { app } = useSisenseContext();

  const renderPlugin = plugins.get(widgetProps.pluginType);
  if (!renderPlugin) {
    return (
      <ErrorBoundaryBox
        error={t('plugins.registerPrompt', { pluginType: widgetProps.pluginType })}
      />
    );
  }

  const dataSource = widgetProps.dataSource ?? app?.defaultDataSource;
  const pluginProps: PluginComponentProps = {
    // TODO - we need to pass dynamically added props, at the same time we wanted to limit amount of the props, so we probably should find a balance here
    ...widgetProps,
    dataSource,
    dataOptions: widgetProps.dataOptions,
    styleOptions: widgetProps.styleOptions,
    filters: widgetProps.filters,
    highlights: widgetProps.highlights,
    description: widgetProps.description,
  };

  return (
    <DynamicSizeContainer defaultSize={getWidgetDefaultSize('line', { hasHeader: true })}>
      <WidgetContainer
        title={widgetProps.title}
        description={widgetProps.description}
        styleOptions={widgetProps.styleOptions as WidgetStyleOptions}
        dataSetName={dataSource && getDataSourceName(dataSource)}
      >
        {renderPlugin(pluginProps)}
      </WidgetContainer>
    </DynamicSizeContainer>
  );
};
