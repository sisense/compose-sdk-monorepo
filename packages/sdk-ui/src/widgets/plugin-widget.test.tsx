import { render } from '@testing-library/react';

import { PluginWidget } from './plugin-widget';
import * as DM from '@/__test-helpers__/sample-ecommerce';

import { usePlugins } from '@/plugins-provider';
import { useSisenseContext } from '@/sisense-context/sisense-context';

vi.mock('@/plugins-provider');
vi.mock('@/sisense-context/sisense-context');

import { Mock } from 'vitest';

describe('PluginWidget', () => {
  const mockUsePlugins = usePlugins as unknown as Mock;
  const mockUseSisenseContext = useSisenseContext as unknown as Mock;

  // We'll store a fake plugin map in memory
  const pluginMap = new Map<string, (props: any) => JSX.Element>();

  beforeEach(() => {
    vi.clearAllMocks();

    // By default, context returns a fallback dataSource
    mockUseSisenseContext.mockReturnValue({
      isInitialized: true,
      app: {
        defaultDataSource: DM.DataSource,
      },
      errorBoundary: {
        showErrorBox: true,
      },
    });

    // By default, no plugins in the map; can add in individual tests
    pluginMap.clear();
    mockUsePlugins.mockReturnValue({
      plugins: {
        get: (key: string) => pluginMap.get(key),
      },
    });
  });

  it('renders an error if plugin is not found', () => {
    const { getByRole } = render(
      <PluginWidget pluginType="unknown-plugin" dataOptions={{}} title={''} />,
    );

    // So let's look for that text in the DOM:
    expect(getByRole('generic', { name: 'error-box' })).toBeInTheDocument();
  });

  it('renders plugin when plugin is found', () => {
    const pluginComponentMock = vi.fn(() => <div>Plugin Content</div>);
    // Add a plugin to the map that just returns some text
    pluginMap.set('my-plugin', pluginComponentMock);

    const { getByText, queryByRole } = render(
      <PluginWidget
        pluginType="my-plugin"
        dataOptions={{ someProp: [] }}
        title={'My awesome plugin'}
        filters={[]}
        highlights={[]}
        styleOptions={{
          someStyle: 'value',
        }}
      />,
    );

    // The error boundary text should not appear
    expect(queryByRole('generic', { name: 'error-box' })).not.toBeInTheDocument();

    // And our plugin component was called
    expect(pluginComponentMock).toHaveBeenCalledOnce();
    expect(pluginComponentMock.mock.calls[0]).toMatchSnapshot();

    // Our plugin text is now rendered
    expect(getByText('Plugin Content')).toBeInTheDocument();
  });
});
