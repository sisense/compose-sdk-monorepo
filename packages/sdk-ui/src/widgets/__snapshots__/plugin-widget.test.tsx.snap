// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PluginWidget > renders plugin when plugin is found 1`] = `
[
  {
    "dataOptions": {
      "someProp": [],
    },
    "dataSource": "Sample ECommerce",
    "description": undefined,
    "filters": [],
    "highlights": [],
    "pluginType": "my-plugin",
    "styleOptions": {
      "someStyle": "value",
    },
    "title": "My awesome plugin",
  },
]
`;
